keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,854 @@
+- There are very few constraints on what a `downscalePodPicker` APIs can be, for example:
+   - they may be fully external to the Kubernetes cluster (e.g. running on an external metrics system)
+   - they may be part of the application which it controls (e.g. a REST API exposed by the application itself)
+   - they may be an adjacent Deployment running on the same Kubernetes cluster (e.g. a Python flask app)
+
+### Risks and Mitigations
+
+- A slow `downscalePodPicker` API may increase the time to downscale a ReplicaSet
+   - MITIGATION: we will provide a `timeoutSeconds` config, if the API has not successfully returned within this time, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- The `downscalePodPicker` may not be reachable at all times, or may return an invalid response
+   - MITIGATION: we will provide a `maxRetries` config, if the API continues to fail after this many attempts, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- Users may attempt to use `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` at the same time, resulting in unexpected behaviour
+   - MITIGATION: documentation will stress that only ONE of `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` should be used per ReplicaSet
+- A malicious (or poorly designed) `downscalePodPicker` API may attempt to overwhelm the controller by returning an extremely large payload
+   - MITIGATION: we will limit the maximum response size to something reasonable, like 1MB
+- Users may find it difficult to build their own `downscalePodPicker` API
+   - MITIGATION: documentation will provide example implementations in common programing languages
+- Users may create non-idempotent `downscalePodPicker` APIs that have side effects (e.g. preventing chosen workers from accepting new tasks):
+   - MITIGATION: documentation will stress that `downscalePodPicker` API must be idempotent, and may be called multiple times
+- Malicious actors may call `downscalePodPicker` APIs from outside the controller loop
+   - MITIGATION: documentation will stress that `downscalePodPicker` APIs should be secured by authentication at the network or request layer
+
+
+## Design Details
+
+#### API Contract:
+
+The contract for a `downscalePodPicker` API will be as follows:
+
+- Request payloads to the API will contain:
+   - `number_of_pods_requested` (`int`): minimum number of Pods to return
+   - `candidate_pods` (`list[string]`): list of Pod-names to choose from
+- Response payloads from the API will contain:
+   - `chosen_pods` (`list[string]`): list of Pod-names chosen to be removed
+   - `tied_pods` (`list[string]`): list of Pod-names we can't decide between
+- Other requirements:
+   - both `chosen_pods` and `tied_pods` can be non-empty
+   - total number of pods returned must be AT LEAST `number_of_pods_requested` (more if there are ties)
```

Comment:
what if the API returns less?

Reply:
I think the logical behavior is to just accept whatever the API gave us, even if `chosen_pods` and `tied_pods` are empty. 

Remember that the API creates a "ranking number" which only has 3 possible values:
- Pods in `chosen_pods` have `rank = 0`
- Pods in `tied_pods` have `rank = 1`
- All remaining Pods have `rank = 3`

There is still a question of what to do if a Pod is returned in both `chosen_pods` and `tied_pods`.

Reply:
Reduced the strength of wording in https://github.com/kubernetes/enhancements/pull/3190/commits/5086ff87ef00561292a2b0d612e3f06a7072451f, to allow less than `number_of_pods_requested` to be returned.

```
total number of pods returned should aim to be AT LEAST `number_of_pods_requested`
```